src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.CopyFile.OSX.cs

```diff
@@ -0,0 +1,117 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO
+{
+    internal static partial class FileSystem
+    {
+        public static partial void CopyFile(string sourceFullPath, string destFullPath, bool overwrite)
```

Comment:
Can't we do this in the pal (`SystemNative_CopyFile`) where we already have different strategies and fallbacks for copying?

Reply:
ok, I assume it's because the function takes paths and not file descriptors.

Reply:
I'm confused what you're saying here, but the reason the last attempt failed was because it had to switch in and out of native mode, I think it's best to keep it in C# unless there's a reason not to.

Reply:
@tmds That was done in the previous attempt which was dismissed. `clonefile` requires paths but that part is easy to propagate. The problematic part is the emulation of file locking which is currently a C# code. I duplicated it in C but that makes the maintenance difficult, and it was ultimately the reason why the previous PR was rejected.

Reply:
When I asked the question, I didn't know yet this function takes string paths.
I answered my own question in the next comment.

Reply:
All good, I just assumed that you would have closed it in that case.